lib/actions-oclif/tag/rm.ts

```diff
@@ -0,0 +1,131 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { flags } from '@oclif/command';
+import { stripIndent } from 'common-tags';
+import Command from '../../command';
+import { ExpectedError } from '../../errors';
+import * as cf from '../../utils/common-flags';
+import { getBalenaSdk } from '../../utils/lazy';
+import { disambiguateReleaseParam } from '../../utils/normalization';
+import { tryAsInteger } from '../../utils/validation';
+
+interface FlagsDef {
+	application?: string;
+	device?: string;
+	release?: string;
+	help: void;
+	app?: string;
+}
+
+interface ArgsDef {
+	tagKey: string;
+}
+
+export default class TagRmCmd extends Command {
+	public static description = stripIndent`
+		Remove a tag from an application, device or release.
+
+		Remove a tag from an application, device or release.
+	`;
+
+	public static examples = [
+		'$ balena tag rm myTagKey --application MyApp',
+		'$ balena tag rm myTagKey --device 7cf02a6',
+		'$ balena tag rm myTagKey --release 1234',
+		'$ balena tag rm myTagKey --release b376b0e544e9429483b656490e5b9443b4349bd6',
+	];
+
+	public static args = [
+		{
+			name: 'tagKey',
+			description: 'the key string of the tag',
+			required: true,
+		},
+	];
+
+	public static usage = 'tag rm <tagKey>';
+
+	public static flags: flags.Input<FlagsDef> = {
+		application: {
+			...cf.application,
+			exclusive: ['app', 'device', 'release'],
+		},
+		device: {
+			...cf.device,
+			exclusive: ['app', 'application', 'release'],
+		},
+		release: {
+			...cf.release,
+			exclusive: ['app', 'application', 'device'],
+		},
+		help: cf.help,
+		app: flags.string({
+			description: "same as '--application'",
+			exclusive: ['application', 'device', 'release'],
+		}),
+	};
+
+	public static authenticated = true;
+
+	public async run() {
+		const { args: params, flags: options } = this.parse<FlagsDef, ArgsDef>(
+			TagRmCmd,
+		);
+
+		// Prefer options.application over options.app
+		options.application = options.application || options.app;
+		delete options.app;
+
+		const balena = getBalenaSdk();
+
+		// Check user has specified one of application/device/release
+		if (!options.application && !options.device && !options.release) {
+			throw new ExpectedError(TagRmCmd.missingResourceMessage);
+		}
+
+		if (options.application) {
+			return balena.models.application.tags.remove(
+				tryAsInteger(options.application),
+				params.tagKey,
+			);
+		}
+		if (options.device) {
+			return balena.models.device.tags.remove(options.device, params.tagKey);
```

Comment:
Ok, though one could argue:
 - ID support for devices is not documented anywhere in CLI help
 - It makes it worse, because with the tryAsInteger hack you can bump into cases where your digit-only short-uuid does not work.

(again we could add `normalization.disambiguateDeviceParam`, but I don't think we should be exposing these internal IDs - just because the SDK supports it, doesn't mean that we should in the CLI - legacy aside).

Reply:
> just because the SDK supports it, doesn't mean that we should in the CLI - legacy aside

This PR is meant to convert the commands from Capitano to oclif, not remove the ability of managing tags via database ID. Also, consider an auto-generated CLI out of the SDK / data model. The fact is that the balena Node SDK, balena Python SDK, balena API all support the "feature" of managing resources via database ID. I am in favour of the feature being removed from the API, SDK and CLI, but as part of a plan discussed in an arch call, and it may require a major CLI release as it would be a breaking change.

> It makes it worse, because with the tryAsInteger hack you can bump into cases where your digit-only short-uuid does not work.

Before this PR, with Capitano, all option arguments that "looked like integer" (say, a 7-char UUID like `0123456`) were automatically converted to number, and would therefore be interpreted by the SDK as a database ID instead of UUID. `tryAsInteger` is a bit more selective by checking whether the string starts with zero: if it does, it is not converted to number, so `tryAsInteger('0123456')` results in a UUID string instead of database ID number. So I don't think that `tryAsInteger` makes it worse. Worse than what? Not worse than before this PR.

A "key feature" of tryAsInteger is that it does not allow strings starting with zero to

Reply:
Yes, I'm not suggesting we remove DB IDs as part of this PR.